### PR TITLE
MB-60945 - Add scorch event for pre merge check

### PR DESCRIPTION
**Pull request by metonymic-smokey**
This PR adds an event to check whether a merge can proceed based on criteria set by the caller(eg. cbft). If not, it's aborted and the merger loop continues its next iteration. 
Also includes minor refactor to the function signature of the event callback registry to accommodate the new event type.

### CODE REVIEW

**Comment by metonymic-smokey** [comment]
TODO Will investigate other metrics for tracking memory used during a merge.

**Comment by metonymic-smokey** [comment]
Potential downsides:
1. Since there is no estimate for the memory used by a merge, it's possible that merges which wouldn't take up very high memory too would be unnecessarily blocked.

File: index/scorch/scorch.go [hunk 1]
```diff
@@ -190,6 +195,14 @@ func (s *Scorch) fireAsyncError(err error) {
 	atomic.AddUint64(&s.stats.TotOnErrors, 1)
 }
 
+func (s *Scorch) continueEvent(kind EventKind) bool {
```
Comment (abhinavdangeti) [inline index/scorch/scorch.go:198]: Should name this better - perhaps `skipEvent`

File: index/scorch/scorch.go [hunk 2]
```diff
@@ -69,8 +69,9 @@ type Scorch struct {
 	rootBolt                 *bolt.DB
 	asyncTasks               sync.WaitGroup
 
-	onEvent      func(event Event)
-	onAsyncError func(err error, path string)
+	onEvent               func(event Event)
+	onAsyncError          func(err error, path string)
+	continueEventCallback func(event Event) bool
```
Comment (abhinavdangeti) [inline index/scorch/scorch.go:74]: skipEventCallback*

File: index/scorch/event.go [hunk 3]
```diff
@@ -24,6 +24,10 @@ var RegistryAsyncErrorCallbacks = map[string]func(error, string){}
 // process init()'ialization.
 var RegistryEventCallbacks = map[string]func(Event){}
 
+// RegistryContinueEventCallback should be treated as read-only after
+// process init()'ialization.
+var RegistryContinueEventCallback = map[string]func(Event) bool{}
```
Comment (abhinavdangeti) [inline index/scorch/event.go:29]: Can we consolidate this map and the above into one - it's ok to change the return type for all the existing events if you must.

**Review by abhinavdangeti — CHANGES_REQUESTED** [review]

**Comment by abhinavdangeti** [comment]
@metonymic-smokey have you checked why the unit tests are failing?

File: index/scorch/event.go [hunk 4]
```diff
@@ -22,7 +22,7 @@ var RegistryAsyncErrorCallbacks = map[string]func(error, string){}
 
 // RegistryEventCallbacks should be treated as read-only after
 // process init()'ialization.
-var RegistryEventCallbacks = map[string]func(Event){}
+var RegistryEventCallbacks = map[string]func(Event) bool{}
```
Comment (abhinavdangeti) [inline index/scorch/event.go:25]: Would you add some commentary above to explain when the expected return values must be true for the function above.

**Review by abhinavdangeti — COMMENTED** [review]
Minor comment, looks ok to me otherwise.

File: index/scorch/event.go (hunk 4 above)
Comment (metonymic-smokey) [reply index/scorch/event.go:25]: The return values are determined by the callbacks. Without a callback, they will fall back to true.

**Review by abhinavdangeti — APPROVED** [review]

**Pull request merged**